### PR TITLE
[2253] Add the possibility of organizing palette tools into sections

### LINKED ISSUE #2253: Let specifiers organize their diagram tools in tool sections in the view DSL

### PR DESCRIPTION
**Pull request by frouene**
Bug: https://github.com/eclipse-sirius/sirius-web/issues/2253

# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [x] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: packages/view/backend/sirius-components-view-diagram-edit/src/main/resources/icons/full/obj16/DiagramToolSection.svg
Comment (pcdavid) [inline packages/view/backend/sirius-components-view-diagram-edit/src/main/resources/icons/full/obj16/DiagramToolSection.svg:1]: Maybe a [folder](https://fonts.google.com/icons?selected=Material+Icons+Outlined:folder:&icon.query=folder&icon.style=Outlined) icon to represent a group or elements?

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/palette/tool-section/ToolSection.tsx [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+const isSingleClickOnDiagramElementTool = (tool: GQLTool): tool is GQLSingleClickOnDiagramElementTool =>
+  tool.__typename === 'SingleClickOnDiagramElementTool';
+
+export const ToolSection = ({ toolSection, onToolClick }: ToolSectionProps) => {
+  const tools = toolSection.tools.filter(isSingleClickOnDiagramElementTool);
+  const initialState = {
+    expanded: false,
+  };
+  const [state, setState] = useState<ToolSectionState>(initialState);
```
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/palette/tool-section/ToolSection.tsx:52]: Do we need a `ToolSectionState` for a single boolean flag?
```js
const [expanded, setExpanded] = useState<boolean>(false);
```

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/palette/Palette.tsx [hunk 2]
```diff
@@ -81,6 +81,24 @@ export const getToolSectionsQuery = gql`
...
+                    __typename
+                    id
+                    label
+                    imageURL
+                    ... on SingleClickOnDiagramElementTool {
+                      targetDescriptions {
+                        id
+                      }
+                      appliesToDiagramRoot
+                      selectionDescriptionId
```
Comment (pcdavid) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/palette/Palette.tsx]: Minor, but maybe define a fragment for this block (reused just above).

**Review by pcdavid — APPROVED** [review]
Possible improvement: EdgeToolSections currently allow the creation of reconnect tools, but these do not normally appear in the palette's UI, so it does not make much sense. It seems they are handled correctly, but it's a little confusing to allow for their creation there.

File: packages/view/backend/sirius-components-view-diagram-edit/src/main/resources/icons/full/obj16/DiagramToolSection.svg
Comment (frouene) [reply packages/view/backend/sirius-components-view-diagram-edit/src/main/resources/icons/full/obj16/DiagramToolSection.svg:1]: OK, I changed it

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/palette/Palette.tsx (hunk 2 above)
Comment (frouene) [reply packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/palette/Palette.tsx]: Done

**Comment by frouene** [comment]
I removed reconnect tools from the EdgeToolSections.

**Review by pcdavid — APPROVED** [review]

**Pull request merged**